x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,332 @@
+    );
+  }
+
+  // While unpacking a tar.gz file with unzipBuffer() will result in a thrown error in the try-catch above,
+  // unpacking a zip file with untarBuffer() just results in nothing.
+  if (paths.length === 0) {
+    throw new PackageInvalidArchiveError(
+      `Uploaded archive seems empty. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+  return paths;
+}
+
+// TODO: everything below performs verification of manifest.yml files, and hence duplicates functionality already implemented in the
+// package registry. At some point this should probably be replaced (or enhanced) with verification based on
+// https://github.com/elastic/package-spec/
+
+function parseAndVerifyArchive(paths: string[]): ArchivePackage {
+  // The top-level directory must match pkgName-pkgVersion, and no other top-level files or directories may be present
+  const toplevelDir = paths[0].split('/')[0];
+  paths.forEach((path) => {
+    if (path.split('/')[0] !== toplevelDir) {
+      throw new PackageInvalidArchiveError('Package contains more than one top-level directory.');
+    }
+  });
+
+  // The package must contain a manifest file ...
+  const manifestFile = `${toplevelDir}/manifest.yml`;
+  const manifestBuffer = cacheGet(manifestFile);
+  if (!paths.includes(manifestFile) || !manifestBuffer) {
+    throw new PackageInvalidArchiveError('Package must contain a top-level manifest.yml file.');
+  }
+
+  // ... which must be valid YAML
+  let manifest;
+  try {
+    manifest = yaml.load(manifestBuffer.toString());
+  } catch (error) {
+    throw new PackageInvalidArchiveError(`Could not parse top-level package manifest: ${error}.`);
+  }
+
+  // Package name and version from the manifest must match those from the toplevel directory
+  const pkgKey = pkgToPkgKey({ name: manifest.name, version: manifest.version });
+  if (toplevelDir !== pkgKey) {
+    throw new PackageInvalidArchiveError(
+      `Name ${manifest.name} and version ${manifest.version} do not match top-level directory ${toplevelDir}`
+    );
+  }
+
+  const { name, version, description, type, categories, format_version: formatVersion } = manifest;
+  // check for mandatory fields
+  if (!(name && version && description && type && categories && formatVersion)) {
+    throw new PackageInvalidArchiveError(
+      'Invalid top-level package manifest: one or more fields missing of name, version, description, type, categories, format_version'
+    );
+  }
+
+  const dataStreams = parseAndVerifyDatasets(paths, name, version);
+  const policyTemplates = parseAndVerifyPolicyTemplates(manifest);
+
+  return {
+    name,
+    version,
+    description,
+    type,
+    categories,
+    format_version: formatVersion,
+    data_streams: dataStreams,
+    policy_templates: policyTemplates,
+  };
+}
+
+function parseAndVerifyDatasets(
```

Comment:
would `parseAndVerifyDataStreams` be more accurate now?

Reply:
Yes.